File: fr/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface User {
  firstName: string;
  lastName: string;
  address: string;
  city: string;
  state: string;
  countryCode: string;
  dob: string;
  age: number;
  gender: string;
  phone: string;
  email: string;
  password: string;
}

@Injectable({
  providedIn: 'root'
})
export class UserService {

  private apiUrl = 'http://localhost:5226/api/Users'; 

  constructor(private http: HttpClient) {}

  getUsers(): Observable<User[]> {
    return this.http.get<User[]>(this.apiUrl);
  }

  addUser(user: User): Observable<any> {
    return this.http.post(this.apiUrl, user);
  }

  updateUser(id: number, user: User): Observable<any> {
    return this.http.put(`${this.apiUrl}/${id}`, user);
  }

  deleteUser(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }
}

File: bk/Service/UserService.cs
using Dapper;
using Microsoft.Extensions.Configuration;
using Npgsql;
using System.Collections.Generic;
using System.Data;
using System.Threading.Tasks;
using WebApplication2.DbContext;
using WebApplication2.Models;

namespace WebApplication2.Service;

public interface IUserService
{
    Task<IEnumerable<User>> GetAllUsersAsync();
    Task<User> GetUserByIdAsync(int id);
    Task<int> CreateUserAsync(User user);
    Task<int> UpdateUserAsync(User user);
    Task<int> DeleteUserAsync(int id);
}

public class UserService(IDbConnectionFactory DbConnectionFactory) : IUserService
{
    

    private readonly IDbConnectionFactory _DbConnectionFactory = DbConnectionFactory;

    
     
    public async Task<IEnumerable<User>> GetAllUsersAsync()
    {
        using var conn = _DbConnectionFactory.CreateConnection();

        string sql = "SELECT * FROM Users";
        return await conn.QueryAsync<User>(sql);
    }

    public async Task<User> GetUserByIdAsync(int id)
    {
        using var conn = _DbConnectionFactory.CreateConnection();
        string sql = "SELECT * FROM Users WHERE Id = @Id";
        return await conn.QueryFirstOrDefaultAsync<User>(sql, new { Id = id });
    }

    public async Task<int> CreateUserAsync(User user)
    {
        using var conn = _DbConnectionFactory.CreateConnection();
        string sql = @"
            INSERT INTO Users 
            (FirstName, LastName, Address, City, State, CountryCode, DOB, Age, Gender, PhoneNumber, Email, Password)
            VALUES 
            (@FirstName, @LastName, @Address, @City, @State, @CountryCode, @DOB, @Age, @Gender, @PhoneNumber, @Email, @Password)";
        return await conn.ExecuteAsync(sql, user);
    }

    public async Task<int> UpdateUserAsync(User user)
    {
        using var conn = _DbConnectionFactory.CreateConnection();
        string sql = @"
            UPDATE Users SET 
                FirstName = @FirstName,
                LastName = @LastName,
                Address = @Address,
                City = @City,
                State = @State,
                CountryCode = @CountryCode,
                DOB = @DOB,
                Age = @Age,
                Gender = @Gender,
                PhoneNumber = @PhoneNumber,
                Email = @Email,
                Password = @Password
            WHERE Id = @Id";
        return await conn.ExecuteAsync(sql, user);
    }

    public async Task<int> DeleteUserAsync(int id)
    {
        using var conn = _DbConnectionFactory.CreateConnection();
        string sql = "DELETE FROM Users WHERE Id = @Id";
        return await conn.ExecuteAsync(sql, new { Id = id });
    }
}